File: core/business-tier/src/main/java/ru/korus/tmis/laboratory/bak/model/OrderInfo.java
package ru.korus.tmis.laboratory.bak.model;

import java.util.LinkedList;
import java.util.List;

/**
 * Author:      Dmitriy E. Nosov <br>
 * Date:        20.11.13, 19:52 <br>
 * Company:     Korus Consulting IT<br>
 * Description:  <br>
 */
public class OrderInfo {
    private String diagnosticCode;
    private String diagnosticName;
    private Priority orderPriority;

    public enum Priority {
        NORMAL, URGENT
    }

    private List<IndicatorMetodic> indicatorList = new LinkedList<IndicatorMetodic>();

    public OrderInfo() {
    }

    public String getDiagnosticCode() {
        return diagnosticCode;
    }

    public void setDiagnosticCode(String diagnosticCode) {
        this.diagnosticCode = diagnosticCode;
    }

    public String getDiagnosticName() {
        return diagnosticName;
    }

    public void setDiagnosticName(String diagnosticName) {
        this.diagnosticName = diagnosticName;
    }

    public Priority getOrderPriority() {
        return orderPriority;
    }

    public void setOrderPriority(Priority orderPriority) {
        this.orderPriority = orderPriority;
    }

    public List<IndicatorMetodic> getIndicatorList() {
        return indicatorList;
    }

    public void addIndicatorList(IndicatorMetodic indicator) {
        this.indicatorList.add(indicator);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("OrderInfo{");
        sb.append("diagnosticCode='").append(diagnosticCode).append('\'');
        sb.append(", diagnosticName='").append(diagnosticName).append('\'');
        sb.append(", orderPriority='").append(orderPriority).append('\'');
        sb.append(", indicatorList=").append(indicatorList);
        sb.append('}');
        return sb.toString();
    }
}

File: core/business-tier/src/main/scala/ru/korus/tmis/laboratory/across/request/OrderInfo.scala
package ru.korus.tmis.laboratory.across.request

import java.util.{List => JList}

import ru.korus.tmis.laboratory.across.{ws => lab2}

import scala.collection.JavaConversions._
import ru.korus.tmis.scala.util.Defaultible
import Defaultible._
import ru.korus.tmis.laboratory.across.ws.Tindicator
import ru.korus.tmis.scala.util.Defaultible

sealed case class OrderInfo(
                             diagnosticCode: Option[String] = None,
                             diagnosticName: Option[String] = None,
                             orderPriority: Option[OrderInfo.OrderPriority] = None,
                             indicators: JList[IndicatorMetodic] = java.util.Collections.emptyList[IndicatorMetodic]
                             )

object OrderInfo {

  object OrderPriority extends Enumeration(1) {
    type OrderPriority = Value
    val Urgent = Value("Urgent")
    val Normal = Value("Normal")
  }

  type OrderPriority = OrderPriority.OrderPriority

  implicit def toLab2OI(v: OrderInfo) = {

    import Utility._

    implicit def locallyDefaultString = setDefault("")
    implicit def locallyDefaultOrderPriority: Defaultible[OrderPriority] = setDefault(OrderPriority.Normal)

    import v._
    val ret = new lab2.OrderInfo
    setAsOptional(diagnosticCode) {
      ret.setDiagnosticCode(_)
    }
    setAsOptional(diagnosticName) {
      ret.setDiagnosticName(_)
    }
    setAsDefaultible(orderPriority) {
      it => ret.setOrderPriority(it.id)
    }

    val indics: JList[lab2.Tindicator] = v.indicators.map {
      IndicatorMetodic.toLab2IM(_)
    }

    val arrayOfIndics = indics.toArray(new Array[Tindicator](indics.size))
    ret.setIndicators(arrayOfIndics)

    ret
  }
}

